File: app/src/main/java/br/com/samueltobias/customercrud/ui/customerlist/CustomerListAdapter.java
package br.com.samueltobias.customercrud.ui.customerlist;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import br.com.samueltobias.customercrud.R;
import br.com.samueltobias.customercrud.model.Customer;
import br.com.samueltobias.customercrud.ui.OnClickListener;

public class CustomerListAdapter extends RecyclerView.Adapter<CustomerListAdapter.ViewHolder> {

    private final Context context;
    private List<Customer> customers = new ArrayList<>();
    private OnClickListener listener;

    CustomerListAdapter(Context context, OnClickListener listener) {
        this.context = context;
        this.listener = listener;
    }

    @NonNull
    @Override
    public CustomerListAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(context).inflate(R.layout.customer_list, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull CustomerListAdapter.ViewHolder holder, final int position) {
        holder.name.setText(customers.get(position).getName());
        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                listener.onClick(position);
            }
        });
    }

    @Override
    public int getItemCount() {
        return customers.size();
    }

    void setCustomers(List<Customer> customers) {
        this.customers = customers;
        notifyDataSetChanged();
    }

    void add(Customer customer) {
        customers.add(customer);
        notifyDataSetChanged();
    }

    List<Customer> getCustomers() {
        return customers;
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        TextView name;

        ViewHolder(@NonNull View itemView) {
            super(itemView);

            name = itemView.findViewById(R.id.customer_list_customer_name);
        }
    }
}

File: app/src/main/java/br/com/samueltobias/customercrud/view/adapter/CustomerListAdapter.kt
package br.com.samueltobias.customercrud.view.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import br.com.samueltobias.customercrud.R
import br.com.samueltobias.customercrud.domain.model.Customer
import br.com.samueltobias.customercrud.view.OnClickListener
import java.util.*

class CustomerListAdapter internal constructor(
        private val context: Context, private val clickListener: OnClickListener
) : RecyclerView.Adapter<CustomerListAdapter.ViewHolder>() {
    private var customers: MutableList<Customer> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(LayoutInflater.from(context).inflate(R.layout.customer_list, parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.customerNameTextView.text = customers[position].name
        holder.itemView.setOnClickListener { clickListener.onClick(position) }
    }

    override fun getItemCount(): Int {
        return customers.size
    }

    fun setCustomers(customers: List<Customer>) {
        this.customers = customers.toMutableList()
        notifyDataSetChanged()
    }

    fun add(customer: Customer) {
        customers.add(customer)
        notifyDataSetChanged()
    }

    fun getCustomers(): List<Customer> {
        return customers
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var customerNameTextView: TextView = itemView.findViewById(R.id.customer_list_customer_name)
    }
}